### PR TITLE
Turn on nullability for Routing

### LINKED ISSUE #24042: Routing public API nullability issue
image

https://github.com/dotnet/aspnetcore/blob/1455aaeff118aeaf2b0ab5f1627c2c8ca5df7f11/src/Http/Routing/src/Patterns/RoutePatternFactory.cs#L58

defaults should allow null

@pranavkm

### PR DESCRIPTION
**Pull request by pranavkm**
We previously only had annotations enabled which resulted in incorrect nullability. This change enables
nullability.

Fixes https://github.com/dotnet/aspnetcore/issues/24042

### CODE REVIEW

File: src/Http/Routing/src/MapRouteRouteBuilderExtensions.cs [hunk 1]
```diff
@@ -154,7 +154,7 @@ public BackCompatInlineConstraintResolver(IInlineConstraintResolver inner, Param
                     return routeConstraint;
                 }
 
-                var parameterPolicy = _parameterPolicyFactory.Create(null, inlineConstraint);
+                var parameterPolicy = _parameterPolicyFactory.Create(null!, inlineConstraint);
```
Comment (JamesNK) [inline src/Http/Routing/src/MapRouteRouteBuilderExtensions.cs:157]: This seems weird. What's going on here?

Comment (pranavkm) [reply src/Http/Routing/src/MapRouteRouteBuilderExtensions.cs:157]: I think I changed the callsite after. This should allow a null.
```suggestion
                var parameterPolicy = _parameterPolicyFactory.Create(null, inlineConstraint);
```

File: src/Http/Http.Abstractions/src/Extensions/EndpointBuilder.cs [hunk 2]
```diff
@@ -14,7 +14,7 @@ public abstract class EndpointBuilder
         /// <summary>
         /// Gets or sets the delegate used to process requests for the endpoint.
         /// </summary>
-        public RequestDelegate? RequestDelegate { get; set; }
+        public RequestDelegate RequestDelegate { get; set; } = default!;
```
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Http/Http.Abstractions/src/Extensions/EndpointBuilder.cs:17]: Nothing enforces it, but I'd assumed an endpoint was always associated with a request delegate. That said, making it nullable is completely safe.

File: src/Http/Routing/src/Constraints/BoolRouteConstraint.cs [hunk 3]
```diff
@@ -14,7 +14,7 @@ public class BoolRouteConstraint : IRouteConstraint
     {
         /// <inheritdoc />
         public bool Match(
-            HttpContext httpContext,
+            HttpContext? httpContext,
             IRouter route,
```
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Http/Routing/src/Constraints/BoolRouteConstraint.cs:18]: IRouter could be null, but routeKey is required by all of the built-in constraints.

File: src/Http/Http.Abstractions/src/Extensions/EndpointBuilder.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Http/Http.Abstractions/src/Extensions/EndpointBuilder.cs:17]: I take it back. `RouteEndpoint` requires that the `RequestDelegate` is non-null. The alternative would be to have RouteEndpointModel throw prior to constructing the endpoint: https://github.com/dotnet/aspnetcore/blob/master/src/Http/Routing/src/RouteEndpointModel.cs#L28. It's weird for RouteEndpointBuilder.Build to throw an arg-null exception

File: src/Http/Routing/src/Matching/CandidateSet.cs [hunk 4]
```diff
@@ -176,15 +177,15 @@ internal static void SetValidity(ref CandidateState candidate, bool value)
...
+        public void ReplaceEndpoint(int index, Endpoint? endpoint, RouteValueDictionary? values)
         {
             // Friendliness for inlining
             if ((uint)index >= Count)
             {
                 ThrowIndexArgumentOutOfRangeException();
             }
             
-            Candidates[index] = new CandidateState(endpoint, values, Candidates[index].Score);
+            Candidates[index] = new CandidateState(endpoint!, values, Candidates[index].Score);
```
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Http/Routing/src/Matching/CandidateSet.cs:188]: Updated CandidateSet

**Review by JamesNK — APPROVED** [review]

**Pull request merged**